librz/analysis/arch/arm/arm_il32.c

```diff
@@ -0,0 +1,2695 @@
+			eff);
+	}
+	return SEQ3(SETL("res0", l), SETL("res1", h), eff);
+}
+
+/**
+ * Capstone: ARM_INS_SADD8, ARM_INS_SHADD8, ARM_INS_SSUB8, ARM_INS_SHSUB8
+ *           ARM_INS_UADD8, ARM_INS_UHADD8, ARM_INS_USUB8, ARM_INS_UHSUB8
+ * ARM: sadd8, shadd8, ssub8, shsub8, uadd8, uhadd8, usub8, uhsub8
+ */
+static RzILOpEffect *sadd8(cs_insn *insn, bool is_thumb) {
+	if (!ISREG(0)) {
+		return NULL;
+	}
+	RzILOpBitVector *a = ARG(1);
+	RzILOpBitVector *b = ARG(2);
+	if (!a || !b) {
+		rz_il_op_pure_free(a);
+		rz_il_op_pure_free(b);
+		return NULL;
+	}
+	RzILOpBitVector *a0 = UNSIGNED(8, a);
+	RzILOpBitVector *a1 = UNSIGNED(8, SHIFTR0(DUP(a), UN(5, 8)));
+	RzILOpBitVector *a2 = UNSIGNED(8, SHIFTR0(DUP(a), UN(5, 16)));
+	RzILOpBitVector *a3 = UNSIGNED(8, SHIFTR0(DUP(a), UN(5, 24)));
+	RzILOpBitVector *b0 = UNSIGNED(8, b);
+	RzILOpBitVector *b1 = UNSIGNED(8, SHIFTR0(DUP(b), UN(5, 8)));
+	RzILOpBitVector *b2 = UNSIGNED(8, SHIFTR0(DUP(b), UN(5, 16)));
+	RzILOpBitVector *b3 = UNSIGNED(8, SHIFTR0(DUP(b), UN(5, 24)));
+	RzILOpBitVector *r0, *r1, *r2, *r3;
+	bool halve = false;
+	switch (insn->id) {
+	case ARM_INS_SHSUB8:
+	case ARM_INS_UHSUB8:
+		halve = true;
+	case ARM_INS_SSUB8:
+	case ARM_INS_USUB8:
+		r0 = SUB(a0, b0);
+		r1 = SUB(a1, b1);
+		r2 = SUB(a2, b2);
+		r3 = SUB(a3, b3);
+		break;
+
+	case ARM_INS_SHADD8:
+	case ARM_INS_UHADD8:
+		halve = true;
+	default: // ARM_INS_SADD8, ARM_INS_UADD8
+		r0 = ADD(a0, b0);
+		r1 = ADD(a1, b1);
+		r2 = ADD(a2, b2);
+		r3 = ADD(a3, b3);
+		break;
+	}
+	bool set_ge = !halve;
+	bool is_signed = insn->id == ARM_INS_SADD8 || insn->id == ARM_INS_SHADD8 || insn->id == ARM_INS_SSUB8 || insn->id == ARM_INS_SHSUB8;
+	if (set_ge) {
+		// Retroactively patch the ops to extend to 8 before the calculation because this is needed for ge
+		// Note: add/sub members here use the same structure, so using just `.add` is fine.
+		RzILOpBitVector *(*cast)(ut32 length, RzILOpBitVector * val) = is_signed ? rz_il_op_new_signed : rz_il_op_new_unsigned;
+		r0->op.add.x = cast(9, r0->op.add.x);
+		r0->op.add.y = cast(9, r0->op.add.y);
+		r1->op.add.x = cast(9, r1->op.add.x);
+		r1->op.add.y = cast(9, r1->op.add.y);
+		r2->op.add.x = cast(9, r2->op.add.x);
+		r2->op.add.y = cast(9, r2->op.add.y);
+		r3->op.add.x = cast(9, r3->op.add.x);
+		r3->op.add.y = cast(9, r3->op.add.y);
+	}
+	RzILOpBitVector *res;
+	if (halve) {
+		res = APPEND(
+			APPEND(
+				SHIFTRA(VARL("res3"), UN(3, 1)),
+				SHIFTRA(VARL("res2"), UN(3, 1))),
+			APPEND(
+				SHIFTRA(VARL("res1"), UN(3, 1)),
+				SHIFTRA(VARL("res0"), UN(3, 1))));
+	} else {
+		res = APPEND(
+			APPEND(
+				UNSIGNED(8, VARL("res3")),
+				UNSIGNED(8, VARL("res2"))),
+			APPEND(
+				UNSIGNED(8, VARL("res1")),
+				UNSIGNED(8, VARL("res0"))));
+	}
+	RzILOpEffect *eff = write_reg(REGID(0), res);
+	if (!eff) {
+		rz_il_op_pure_free(r0);
+		rz_il_op_pure_free(r1);
+		rz_il_op_pure_free(r2);
+		rz_il_op_pure_free(r3);
+		return NULL;
+	}
+	if (set_ge) {
+		eff = SEQ2(
```

Comment:
Is it possible to simplify this somehow?

Reply:
Actually yes it can, good point.